modules/transport-netty4/src/main/java/org/elasticsearch/transport/Netty4Plugin.java

```diff
@@ -66,11 +74,28 @@ public Settings additionalSettings() {
                 .build();
     }
 
-    public void onModule(NetworkModule networkModule) {
-        if (networkModule.canRegisterHttpExtensions()) {
-            networkModule.registerHttpTransport(NETTY_HTTP_TRANSPORT_NAME, Netty4HttpServerTransport.class);
-        }
-        networkModule.registerTransport(NETTY_TRANSPORT_NAME, Netty4Transport.class);
+    @Override
+    public List<TransportFactory<Transport>> getTransportFactory() {
+        return Collections.singletonList(new TransportFactory<Transport>(NETTY_TRANSPORT_NAME) {
+            @Override
+            public Transport createTransport(Settings settings, ThreadPool threadPool, BigArrays bigArrays,
+                                             CircuitBreakerService circuitBreakerService, NamedWriteableRegistry namedWriteableRegistry,
+                                             NetworkService networkService) {
+                return new Netty4Transport(settings, threadPool, networkService, bigArrays, namedWriteableRegistry, circuitBreakerService);
+            }
+        });
+    }
+
+    @Override
+    public List<TransportFactory<HttpServerTransport>> getHttpTransportFactory() {
+        return Collections.singletonList(new TransportFactory<HttpServerTransport>(NETTY_HTTP_TRANSPORT_NAME) {
```

Comment:
This way of declaring the factory is fine I guess. Maybe I rescind my comment about returning a `Map<String, Factory>` in the plugin. This is just fine. A map is more normal for our plugins I think because we usually *can* use an inline function to implement the extension. And it makes it blindingly obvious that you can't return the same name twice.